insights/client/crypto.py

```diff
@@ -0,0 +1,270 @@
+
+        # Older systems do not support `gpgconf --kill`.
+        # The socket may remove its socket file after `rmtree()` has determined
+        # it should be deleted, but before the actual deletion occurs.
+        # This would cause a FileNotFoundError/OSError.
+        for _ in range(5):
+            try:
+                shutil.rmtree(self._home)
+                break
+            except OSError as exc:
+                if exc.errno == errno.ENOENT:
+                    # The file has already been removed by the `gpg-agent`.
+                    continue
+                raise
+        else:
+            logger.debug(
+                "could not clean up temporary GPG directory '{path}'".format(path=self._home)
+            )
+
+    def _run(self, command):
+        """Run the actual command.
+
+        :returns: The result of the shell command.
+        :rtype: GPGCommandResult
+        """
+        self._raw_command = ["/usr/bin/gpg", "--homedir", self._home] + command
+        process = subprocess.Popen(
+            self._raw_command,
+            stdout=subprocess.PIPE, stderr=subprocess.PIPE,
+        )
+        stdout, stderr = process.communicate()
+
+        result = GPGCommandResult(
+            ok=process.returncode == 0,
+            return_code=process.returncode,
+            stdout=stdout.decode("utf-8"),
+            stderr=stderr.decode("utf-8"),
+            command=self,
+        )
+
+        if result.ok:
+            logger.debug("gpg command {command}: ok".format(command=command))
+        else:
+            logger.debug(
+                "gpg command {command} returned non-zero code: {result}".format(
+                    command=command, result=result,
+                )
+            )
+
+        return result
+
+    def evaluate(self):
+        """Run the command.
+
+        :returns: The result of the shell command.
+        :rtype: GPGCommandResult
+        """
+        try:
+            result = self._setup()
+            if not result.ok:
+                logger.debug("gpg setup failed")
+                return result
+
+            logger.debug("running gpg in the temporary environment")
+            return self._run(self.command)
+        finally:
+            self._cleanup()
+
+
+def verify_gpg_signed_file(file, signature, key):
+    """
+    Verify a file that was signed using GPG.
+
+    :param file: A path to the signed file.
+    :type file: str
+    :param signature: A path to the detached signature.
+    :type signature: str
+    :param key: Path to the public GPG key on the filesystem to check against.
+    :type key: str
+
+    :returns: Evaluated GPG command.
+    :rtype: GPGCommandResult
+    """
+    if not os.path.isfile(file):
+        logger.debug(
+            "cannot verify signature of '{file}', file does not exist".format(
+                file=file
+            )
+        )
+        raise OSError(errno.ENOENT, "File '{file}' not found".format(file=file))
```

Comment:
I wonder whether here, rather than raising an exception, `GPGCommandResult(false)` should be returned

according to the code where now `verify_gpg_signed_file()` is used, this situation used to result in `/usr/bin/gpg` failing, and thus handled as failure for it; now this will raise an exception, and I'm not sure it will be handled well within the rest of the core code...

in case this is changed, please add a test case for this situation, together with the newly added ones